### PR TITLE
Support chat conversation for StaticLLMPipeline

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
# Overview

Adding chat mode support for `StaticLLMPipeline`. 

The current implementation is naive - aggregates the entire chat conversation and pass as new prompt on every new `generate` call.

### CODE REVIEW

File: src/cpp/src/llm_pipeline_static.cpp [hunk 1]
```diff
@@ -127,6 +126,7 @@ StaticLLMPipeline::StaticLLMPipeline(
     // (1) Read the template model - this will be kvcache model
     auto kvcache_model = core.read_model(path / "openvino_model.xml");
     // (2) TODO: Expose KV-cache input and output layers from kvcache model
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Done

File: src/cpp/src/llm_pipeline_static.cpp [hunk 2]
```diff
@@ -245,22 +267,25 @@ EncodedResults StaticLLMPipeline::generate(
     ov::genai::EncodedResults results;
     // NB: Only batch=1 is supported now
     results.scores.resize(1u);
+    results.scores[0] = 0u;
```
Comment (TolyaTalamanov) [inline src/cpp/src/llm_pipeline_static.cpp:270]: Was missed previously

File: src/cpp/src/llm_pipeline_static.cpp [hunk 3]
```diff
@@ -271,13 +296,13 @@ EncodedResults StaticLLMPipeline::generate(
     // NB: Now there are prompt_len tokens in KV-cache
     m_kvcache_desc.num_stored_tokens += prompt_len;
     int64_t last_token = utils::argmax(m_prefill_request.get_tensor("logits"), 0);
+    results.tokens[0].push_back(last_token);
```
Comment (TolyaTalamanov) [inline src/cpp/src/llm_pipeline_static.cpp:299]: also was missed

File: src/cpp/src/llm_pipeline_static.cpp [hunk 4]
```diff
@@ -309,13 +334,12 @@ EncodedResults StaticLLMPipeline::generate(
 
         last_token = utils::argmax(m_kvcache_request.get_tensor("logits"), 0);
         results.tokens[0].push_back(last_token);
-        results.scores[0] = 0u;
```
Comment (TolyaTalamanov) [inline src/cpp/src/llm_pipeline_static.cpp:312]: has no sense in loop

File: samples/cpp/chat_sample/chat_sample.cpp [hunk 5]
```diff
@@ -4,15 +4,14 @@
 #include "openvino/genai/llm_pipeline.hpp"
 
 int main(int argc, char* argv[]) try {
-    if (2 != argc) {
-        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR>");
+    if (2 != argc && 3 != argc) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> <DEVICE>");
```
Comment (slyalin) [inline samples/cpp/chat_sample/chat_sample.cpp:8]: Show that `DEVICE` is optional

Comment (TolyaTalamanov) [reply samples/cpp/chat_sample/chat_sample.cpp]: Sure, makes sense, done

File: samples/cpp/chat_sample/chat_sample.cpp [hunk 6]
```diff
@@ -4,15 +4,14 @@
 #include "openvino/genai/llm_pipeline.hpp"
 
 int main(int argc, char* argv[]) try {
-    if (2 != argc) {
-        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR>");
+    if (2 != argc && 3 != argc) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> <DEVICE>(optional)");
```
Comment (Wovchena) [inline samples/cpp/chat_sample/chat_sample.cpp:8]: Don't change that. The idea is to minimize the number of cmd args. You can extend the comment instead with NPU

File: src/cpp/src/llm_pipeline_static.cpp [hunk 7]
```diff
@@ -77,18 +77,15 @@ void reshape_to_static(std::shared_ptr<ov::Model> model,
     model->reshape(new_shapes);
 }
 
-void fill_tensor(ov::Tensor tensor, int64_t fill_val) {
+void fill_tensor(ov::Tensor tensor, int64_t fill_val, int32_t offset = 0) {
```
Comment (Wovchena) [inline src/cpp/src/llm_pipeline_static.cpp:80]: Why not `size_t`. Or `ptrdiff_t` if you expect a negative offset, which doesn't make sense for `tensor_data + offset`? They are for pointer arithmetic.

File: src/cpp/src/llm_pipeline_static.cpp [hunk 8]
```diff
@@ -198,9 +206,23 @@ DecodedResults StaticLLMPipeline::generate(
...
+    if (m_is_chat_conversation) {
+        m_history.push_back({{"role", "user"}, {"content", prompt}});
+        constexpr bool add_generation_prompt = true;
+        prompt = m_tokenizer.apply_chat_template(m_history, add_generation_prompt);
+    }
+
+    auto tokenized_input = m_tokenizer.encode(prompt);
     auto encoded_results = generate(tokenized_input, config, streamer);
-    return {m_tokenizer.decode(encoded_results.tokens), encoded_results.scores};
+    DecodedResults decoded_results = {m_tokenizer.decode(encoded_results.tokens), encoded_results.scores};
```
Comment (Wovchena) [inline src/cpp/src/llm_pipeline_static.cpp:219]: Just to make it clear. Does that implementation runs the whole history on every generate()?

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: samples/cpp/chat_sample/chat_sample.cpp (hunk 6 above)
Comment (TolyaTalamanov) [reply samples/cpp/chat_sample/chat_sample.cpp]: Returned back anyway

File: src/cpp/src/llm_pipeline_static.cpp (hunk 8 above)
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp:219]: Yes, 
```
EncodedResults StaticLLMPipeline::generate(
    const EncodedInputs& inputs,
    OptionalGenerationConfig generation_config,
    StreamerVariant streamer)
```
knows nothing about chat conversation, it consumes input prompt up to 1024 tokens

File: src/cpp/src/llm_pipeline_static.cpp (hunk 7 above)
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Overlooked it, thanks, fixed

File: src/cpp/src/llm_pipeline_static.cpp (hunk 8 above)
Comment (dmatveev) [reply src/cpp/src/llm_pipeline_static.cpp:219]: Why only 1024? Shouldn't this be defined by the model? Also, when we turn the model to static shape form, what we use as a guidance to fix the input prompt/kvcache size?

Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp:219]: > Shouldn't this be defined by the model?
It shouldn't model has dynamic shape

**Review by Wovchena — APPROVED** [review]

File: src/cpp/src/llm_pipeline_static.cpp [hunk 9]
```diff
@@ -179,6 +176,17 @@ StaticLLMPipeline::StaticLLMPipeline(
 ) : StaticLLMPipeline(path, path.string(), device, config) {
 }
 
+void StaticLLMPipeline::start_chat(const std::string& system_message) {
+    if (system_message.empty()) {
+        m_history.push_back({{"role", "system"}, {"content", system_message}});
+    }
+    m_is_chat_conversation = true;
+};
+void StaticLLMPipeline::finish_chat() {
```
Comment (pavel-esir) [inline src/cpp/src/llm_pipeline_static.cpp:185]:
```suggestion

void StaticLLMPipeline::finish_chat() {
```

File: src/cpp/src/llm_pipeline_static.cpp [hunk 10]
```diff
@@ -179,6 +176,17 @@ StaticLLMPipeline::StaticLLMPipeline(
 ) : StaticLLMPipeline(path, path.string(), device, config) {
 }
 
+void StaticLLMPipeline::start_chat(const std::string& system_message) {
+    if (system_message.empty()) {
```
Comment (pavel-esir) [inline src/cpp/src/llm_pipeline_static.cpp:180]: Looks like here is a typo, system message should be added to history only if not empty.

```suggestion
    if (!system_message.empty()) {
```

**Review by pavel-esir — APPROVED** [review]
I think there was a typo, but overall looks good. Approved

File: src/cpp/src/llm_pipeline_static.cpp (hunk 9 above)
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp:185]: What is the difference?

Comment (pavel-esir) [reply src/cpp/src/llm_pipeline_static.cpp:185]: just formatting, for readability a blank line between function definitions

File: src/cpp/src/llm_pipeline_static.cpp [hunk 11]
```diff
@@ -179,6 +176,18 @@ StaticLLMPipeline::StaticLLMPipeline(
 ) : StaticLLMPipeline(path, path.string(), device, config) {
 }
 
+void StaticLLMPipeline::start_chat(const std::string& system_message) {
```
Comment (Wovchena) [inline src/cpp/src/llm_pipeline_static.cpp:179]: I just noticed. After this is merged you may want to add `override` keyword

Comment (Wovchena) [reply src/cpp/src/llm_pipeline_static.cpp:179]: Never mind

**Pull request merged**